File: src/Julia/Preprocessing/findexec.jl
using Gencode

function findexec(filename, version)

status = Sys.which(filename);
if status != nothing
    return filename;
end
status = Sys.which("/usr/bin/" * filename);
if status != nothing
    filename = "/usr/bin/" * filename;
    return filename;
end
status = Sys.which("/usr/local/bin/" * filename);
if status != nothing
    filename = "/usr/local/bin/" * filename;
    return filename;
end
status = Sys.which("/opt/local/bin/" * filename);
if status != nothing
    filename = "/opt/local/bin/" * filename;
    return filename;
end
if status == nothing
    print("Exasim could not find " * filename * " in /usr/bin, /usr/local/bin, /opt/local/bin\n");
    if Sys.isapple()
        print("Exasim tries to find " * filename * " in /Applications. It may take a while.\n");
        a = read(Gencode.string2cmd("find /Applications -name " * filename * " -type f"), String);
        if length(a)>0
            ii = findlast("/MacOS/" * filename, a);
            newfilename = a[1:ii[end]];
            print("Exasim found " * filename * " at " * newfilename * "\n");
            print("Please open initializepde.jl in the folder Exasim/" * version * "/Julia/Preprocessing\n");
            qstr1 = """ "$filename" """;
            qstr2 = """ "$newfilename" """;
            print("Replace pde." * filename * " =" * qstr1 * " with pde." * filename * " =" * qstr2 * "\n");
            print("Doing so will prevent Exasim from searching " * filename * " again to save time.\n");
            print("Read the above instructions carefully and press any key to continue...\n");
            sleep(20);
            filename = newfilename;
            return filename;
        end
    # elseif Sys.isunix()
    # elseif Sys.windows()
    end
    mystr = "Exasim could not find " * filename * " on your system.\n";
    mystr = mystr * "Please see the documentation to install " * filename * ".\n"
    mystr = mystr * "After installing " * filename * ", open initializepde.jl in the folder Exasim/" * version * "/Julia/Preprocessing\n";
    qstr1 = """ "$filename" """;
    newfilename = "/path/to/executable/" * filename;
    qstr2 = """ "$newfilename" """;
    mystr = mystr * "and replace pde." * filename * " =" * qstr1 * "with pde." * filename * " =" * qstr2 * "\n";
    print(mystr);
    error("Exit Exasim!")
end

end

File: Version0.2/Matlab/Preprocessing/findexec.m
function filename = findexec(filename, version)

[status,~] = system("which " + filename);
if status==0
    return;
end
[status,~] = system("which /usr/bin/" + filename);
if status==0
    filename = "/usr/bin/" + filename;
    return;
end
[status,~] = system("which /usr/local/bin/" + filename);
if status==0
    filename = "/usr/local/bin/" + filename;
    return;
end
[status,~] = system("which /opt/local/bin/" + filename);
if status==0
    filename = "/opt/local/bin/" + filename;
    return;
end
if status~=0
    disp("Exasim could not find " + filename + " in /usr/bin, /usr/local/bin, /opt/local/bin");
    if ismac
        disp("Exasim tries to find " + filename + " in /Applications. It may take a while.");
        [~, b] = system("find /Applications -name " + filename + " -type f");
        if ~isempty(b)
            ii = strfind(b, "/MacOS/" + filename);
            newfilename = b(1:ii(1)) + "MacOS/" + filename;
            disp("Exasim found " + filename + " at " + newfilename);
            disp("Please open initializepde.m in the folder Exasim/" + version + "/Matlab/Preprocessing");
            disp("Replace pde." + filename + " = " + """" + filename + """" + " with pde." + filename + " = " + """" + newfilename + """");
            disp("Doing so will prevent Exasim from searching " + filename + " again");
            disp("Read the above instructions carefully and press any key to continue...");
            open("initializepde.m");
            pause();
            filename = newfilename;
            return;
        end
%     elseif isunix
%     elseif ispc
    end
    disp("Exasim could not find " + filename + " on your system.");
    disp("Please see the documentation to install " + filename + ".");
    disp("After installing " + filename + ", open initializepde.m in the folder Exasim/Version0.1/Matlab/Preprocessing");
    disp("and replace pde." + filename + " = " + """" + filename + """" + " with pde." + filename + " = " + """" + "/path/to/executable/" + filename + """");
    error("Exit Exasim!");
end

end
